File: greed/SRM 710/ReverseMancala.cpp
#include "bits/stdc++.h"
using namespace std;

class ReverseMancala {
    private:
    void processA(vector<int> &arr, const unsigned int pos){
        assert(pos<arr.size());
        int foo=arr[pos];
        arr[pos]=0;
        for(unsigned int i=0; i<arr.size(); i++){
            arr[i]+=foo/arr.size();
        }
        for(unsigned int i=1; i<=foo%arr.size(); i++){
            arr[(pos+i)%arr.size()]++;
        }
    }
    int toLast(vector<int> &arr, int *last=NULL){
        for(unsigned int i=0; i<arr.size()-1; i++){
            if(arr[i]>0){
                if(last!=NULL){
                    *last=(i+arr[i])%arr.size();
                }
                this->processA(arr, i);
                return i;
            }
        }
        return -1;
    }
    public:
    vector<int> findMoves(vector<int> start, vector<int> target){
        vector<int> ans;
        while(true){
            int foo=this->toLast(start);
            if(!~foo)break;
            ans.push_back(foo);
        }
        stack<int> seq;
        while(true){
            int last;
            int foo=this->toLast(target, &last);
            if(!~foo)break;
            seq.push(last);
        }
        for(; !seq.empty(); seq.pop()){
            ans.push_back(start.size()+seq.top());
        }
        return ans;
    }
};

File: greed/SRM 710/ReverseMancala.py
arr = [int(i) for i in raw_input().split()]
while True:
    c = int(raw_input())
    if c<len(arr):
        foo = arr[c]
        arr[c] = 0
        while foo>0:
            c = (c+1)%len(arr)
            foo -= 1
            arr[c] += 1
    else:
        c -= len(arr)
        foo = 0
        while arr[c]>0:
            foo += 1
            arr[c] -= 1
            c = (c-1)%len(arr)
        arr[c] += foo
    print arr
